CONTRIBUTING.md

```diff
@@ -84,6 +84,21 @@ only on `nightly` toolchain. If you can, test both toolchains.
 * Push your commits to GitHub and create a pull request against Cargo's
 `master` branch.
 
+### Running tests on Windows
+
+To ensure all of Cargo's tests pass on Windows, there are a few things to
+verify with your environment. Besides having a normal Rust development
+environment (with CMake installed), check the following:
+
+* `sh` from MSYS should be in your PATH (ahead of any other `sh` installations
+  such as Cygwin).  If you install Git for Windows, choose the option `Use Git
+  and optional Unix tools from the Windows Command Prompt` to use the shell
+  that is bundled with Git, or place the path `C:\Program Files\Git\usr\bin`
+  towards the top of your PATH list.
+* Avoid using a `~/.cargo/config` file.  Some tests run in a temp directory
```

Comment:
This seems like a bug I think we should be able to fix, right?

Reply:
If you have any ideas how to fix it, I could try to make a PR.  I don't know of a way to prevent cargo from recursing up the directory structure.  And I don't know of a way to get a Temp directory in Windows that is not under %USERPROFILE%.  (See #4601)

Reply:
We can just override the env var, right?

Reply:
We could override TMP, I just don't know of a safe path to use.  It needs be to outside of the cargo repo, and outside of the user's home directory.  We could create a random directory in the root of a drive (I think), but that doesn't sound good to me.

Reply:
Cargo creates a home/tmp folder for each test, I think that can be used?

Reply:
If [this](https://github.com/rust-lang/cargo/blob/2f779743f4579febf674dbbb6e24bb2e709a7982/tests/cargotest/support/paths.rs#L54-L56) is what you are referring to, the directories cargotest creates are all inside the `target` directory of the cargo checkout.  The intent of these particular tests is to run outside of that directory hierarchy to avoid picking up any `.cargo/config` files.  I believe you wrote the original code [here](https://github.com/rust-lang/cargo/commit/f4e91546a44a393000dea8607920a5cee52a1594).

Reply:
I think we can use the `tempdir` crate then? Or anything that's outside?

Reply:
It is using `tempdir`, that's the problem.  `tempdir` uses the temp directory on Windows which is `~/AppData/Local/Temp/`.  Unfortunately since that is under the user's home directory, cargo will read `~/.cargo/config`, and if you have `[cargo-new]` entries, the test will fail since those take priority.  The only fixes I know of are:
* Create a temp directory somewhere safe.  But where?
* Somehow prevent cargo from recursing up the directory structure.

Reply:
Sounds like we should add a way for Cargo to avoid traversing upwards in tests?

Reply:
Hm, I just discovered #3869 which added `__CARGO_TEST_ROOT` as a hack to prevent this very issue.  I will investigate if this can be leveraged to make this work, it should be simple.

Reply:
OK, opened #4630 with a proposed fix.

Reply:
Thanks!